keps/sig-node/20190226-pod-overhead.md

```diff
@@ -98,152 +125,119 @@ Introduce a Pod.Spec.Resources field on the pod to specify the pods overhead.
+left out of the scope of this proposal.
 
-For scheduling, the pod resource requests are added to the container resource requests.
+For scheduling, the pod `Overhead` resource requests are added to the container resource requests.
 
 We don't currently enforce resource limits on the pod cgroup, but this becomes feasible once
 pod overhead is accountable. If the pod specifies a resource limit, and all containers in the
 pod specify a limit, then the sum of those limits becomes a pod-level limit, enforced through the
 pod cgroup.
 
-Users are not expected to manually set the pod resources; if a runtimeClass is being utilized,
-the manual value will be discarded. See RuntimeController for the proposal for setting these
-resources.
+Users are not expected to manually set `Overhead`; any prior values will being set will result in the workload
+being rejected. If runtimeClass is configured and selected in the PodSpec, `Overhead` will be set to the value
+defined in the corresponding runtimeClass. This is described in detail in
+[RuntimeClass admission controller](#runtimeclass-admission-controller).
+
+Being able to specify resource requirements for a workload at the pod level instead of container
+level has been discussed, but isn't proposed in this KEP.
 
-Being able to specify resource requirements for a workload at the pod level instead of container level
-has been discussed, but isn't proposed in this KEP.
+In the event that pod-level requirements are introduced, pod overhead should be kept separate. This
+simplifies several scenarios:
+ - overhead, once added to the spec, stays with the workload, even if runtimeClass is redefined
+ or removed.
+ - the pod spec can be referenced directly from scheduler, resourceQuote controller and kubelet,
+ instead of referencing a runtimeClass object which could have possibly been removed.
 
-In the event that pod-level requirements are introduced, pod overhead should be kept separate. This simplifies
-several scenarios:
- - overhead, once added to the spec, stays with the workload, even if runtimeClass is redefined or removed.
- - the pod spec can be referenced directly from scheduler, resourceQuote controller and kubelet, instead of referencing
- a runtimeClass object which could have possibly been removed.
+#### Container Runtime Interface (CRI)
 
+The pod cgroup is managed by the Kubelet, so passing the pod-level resource to the CRI implementation
+is not strictly necessary. However, some runtimes may wish to take advantage of this information, for
+instance for sizing the Kata Container VM.
+
+The LinuxContainerResources message is added to the LinuxPodSandboxConfig, as an optional field. Unlike
+the Resources field on the Kubernetes PodSpec, the resources field in the LinuxPodSandboxConfig matches
+the pod-level limits (i.e. the total of pod overhead & container limits). The field is only provided for
+pods where each container provides limits (ie, guaranteed pods and a subset of burstable pods).
```

Comment:
How does Kata handle those other pods? Would having the pod overhead still be useful in those situations?

Reply:
I think it could be useful to make this information available in either situation.

If requests and limits are not provided, the sizing of the virtual machine is based on kata defaults (default configurable parameters for Kata, which are currently a single vCPU and 2048 MB memory).

Today, we adjust the VM sizing each time we receive a container spec, and only if limits and / or requests are provided.  Receiving this field, included overhead, will help the runtime know an appropriate initial size up front (at VM creation time), potentially avoiding needing to read each container spec and hotplugging CPUs.

In the scenario where no req/limits are defined, are there any default constraints for the pod cgroup?